### PR TITLE
 boards: arm: Add support for Silabs Happy Gecko SLSTK3400A board

### PR DESCRIPTION
**Pull request by mtuxpe**
The EFM32 Happy Gecko USB-enabled kit contains sensors and
peripherals demonstrating some of the EFM32HG MCU capabilities.
This includes:
* the Gecko SDK files for the EFM32HG SoCs
* the emlib peripheral driver files

Origin: Silicon Labs Gecko SDK
URL: https://github.com/SiliconLabs/Gecko_SDK
Version: v5.1.2 (SHA: 938464c68e6c3b2237388a692f767bb0767ec010)
Purpose: Add support for Silicon Labs EXX32 SoCs
License: Zlib
Maintained-by: External

To update the Gecko SDK, download the current version from the given URL
and replace the following folders:
* Gecko_SDK/platform/Device/ should replace ext/hal/silabs/gecko/Device/
* Gecko_SDK/platform/emlib/ should replace ext/hal/silabs/gecko/emlib/

Signed-off-by: Marcio Montenegro <mtuxpe@gmail.com>

### CODE REVIEW

**Comment by galak** [comment]
Can you merge the commits down so they are clean?

**Comment by galak** [comment]
Look into git rebase and the ability to 'squash' commits

https://blog.carbonfive.com/2017/08/28/always-squash-and-rebase-your-git-commits/

File: boards/arm/efm32hg_slstk3400a/doc/efm32hg_slstk3400a.rst [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+********
+
+The EFM32 Happy Gecko Starter Kit EFM32WG-STK3800 contains a MCU from the
+EFM32HG family built on ARM® Cortex®-M4F processor with excellent low
+power capabilities.
+
+.. image:: happy-gecko-starter-kit.jpg
+   :width: 375px
+   :align: center
+   :alt: EFM32HG-SLSTK3400A
```
Comment (chrta) [inline boards/arm/efm32hg_slstk3400a/doc/efm32hg_slstk3400a.rst]: If you took the image from Silicon Labs, you should credit them, see https://www.silabs.com/community/general-interest/forum.topic.html/usage_of_board_pictu-ydFZ

I did this there: https://github.com/zephyrproject-rtos/zephyr/pull/6841/commits/3092d6aaf5c864601bae42f8fd0c2abd95521047#diff-60f73521f2a29f145ead3733a9f66c26R13

File: boards/arm/efm32hg_slstk3400a/doc/efm32hg_slstk3400a.txt [hunk 2]
```diff
@@ -0,0 +1,175 @@
+.. _efm32wg_stk3800:
```
Comment (chrta) [inline boards/arm/efm32hg_slstk3400a/doc/efm32hg_slstk3400a.txt]: This file should be removed (duplication of the rst file)

**Comment by mtuxpe** [comment]
Thanks for review Christian

File: boards/arm/efm32hg_slstk3400a/doc/efm32hg_slstk3400a.rst [hunk 3]
```diff
@@ -0,0 +1,177 @@
+.. _efm32hg_slstk3400a:
+
+EFM32HG-SLSTK3400A
+###############
```
Comment (dbkinder) [inline boards/arm/efm32hg_slstk3400a/doc/efm32hg_slstk3400a.rst]: Title underline must be at least as long as the title, add some more ####'s

File: boards/arm/efm32hg_slstk3400a/doc/efm32hg_slstk3400a.rst [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+- ARM Cortex M0+ with 64 kB Flash and 8 kB RAM
+- 128 X 128 pixel Memory LCD 
+- 2 user buttons, 2 user LEDs and 2 touch buttons
+- 20 pin expansion header
+- Silicon Labs Si7021 Relative Humidity/Temperature sensor
+- USB device interface
+- Integrated SEGGER J-Link USB debugger/emulator with debug out functionality
+
+
+For more information about the EFM32HG SoC and EFM32HG-SLSTK3400A Happy Gecko Starter Kit:
```
Comment (dbkinder) [inline boards/arm/efm32hg_slstk3400a/doc/efm32hg_slstk3400a.rst]: ``See these documents for more information ...``

File: boards/arm/efm32hg_slstk3400a/doc/efm32hg_slstk3400a.rst [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+The default configuration can be found in the defconfig file:
+
+	``boards/arm/efm32hg_slstk3400a/efm32hg_slstk3400a_defconfig``
+
+Other hardware features are currently not supported by the port.
+
+Connections and IOs
+===================
+
+The EFM32HG SoC has six gpio controllers (PORTA to PORTF), but only three are
```
Comment (dbkinder) [inline boards/arm/efm32hg_slstk3400a/doc/efm32hg_slstk3400a.rst]: ``GPIO``

File: boards/arm/efm32hg_slstk3400a/doc/efm32hg_slstk3400a.rst [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+The EFM32WG-STK3800 includes an `J-Link`_ serial and debug adaptor built into the
+board. The adaptor provides:
+
+- A USB connection to the host computer, which exposes a Mass Storage and a
+  USB Serial Port.
+- A Serial Flash device, which implements the USB flash disk file storage.
+- A physical UART connection which is relayed over interface USB Serial port.
+
+Flashing an application to EFM32-STK3800A
+----------------------------------------
```
Comment (dbkinder) [inline boards/arm/efm32hg_slstk3400a/doc/efm32hg_slstk3400a.rst]: Title underline must be at least as long as the title, add one more ``-``

File: boards/arm/efm32hg_slstk3400a/doc/efm32hg_slstk3400a.rst [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+The sample application :ref:`hello_world` is used for this example.
+Build the Zephyr kernel and application:
+
+.. zephyr-app-commands::
+   :zephyr-app: samples/hello_world
+   :board: efm32hg_slstk3400a
+   :goals: build
+
+Connect the EFM32HG-SLSTK3800A to your host computer using the USB port and you
+should see a USB connection which exposes a Mass Storage (STK3400) and a
```
Comment (dbkinder) [inline boards/arm/efm32hg_slstk3400a/doc/efm32hg_slstk3400a.rst]: ``that exposes a mass storage device``

File: boards/arm/efm32hg_slstk3400a/doc/efm32hg_slstk3400a.rst [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+USB Serial Port. Copy the generated zephyr.bin in the STK3400 drive.
+
+Open a serial terminal (minicom, putty, etc.) with the following settings:
+
+- Speed: 115200
+- Data: 8 bits
+- Parity: None
+- Stop bits: 1
+
+Reset the board and you should be able to see on the corresponding Serial Port
```
Comment (dbkinder) [inline boards/arm/efm32hg_slstk3400a/doc/efm32hg_slstk3400a.rst]: Reset the board and you will see this message written to the serial port:

**Review by dbkinder — CHANGES_REQUESTED** [review]

**Review by dbkinder — APPROVED** [review]
Thanks!

**Review by galak — CHANGES_REQUESTED** [review]
Can you add a README per: http://docs.zephyrproject.org/contribute/contribute_non-apache.html
for the SDK import.

**Comment by mtuxpe** [comment]
Hello Kumar,
First sorry for delay.There is a README file on folder ext/hal/silabs/gecko

https://github.com/zephyrproject-rtos/zephyr/tree/master/ext/hal/silabs/gecko

So I need to replace the current file to the new format?

Best regards,

Marcio

**Comment by galak** [comment]
Yes please do (try and keep the info about updating & folders used from the HAL in the new README).  As new EXT contributions need to get TSC & zephyr board review its easiest to start with having something in the README format described in the contribution guidelines.

File: drivers/serial/uart_gecko.c [hunk 9]
```diff
@@ -329,8 +329,13 @@ static void uart_gecko_config_func_1(struct device *dev);
 #endif
 
 static const struct uart_gecko_config uart_gecko_1_config = {
-	.base = UART1,
+#ifdef CONFIG_SOC_PART_NUMBER_EFM32HG322F64
```
Comment (diegosueiro) [inline drivers/serial/uart_gecko.c]: I think you can use "#ifdef CONFIG_SOC_SERIES_EFM32HG" instead of "#ifdef CONFIG_SOC_PART_NUMBER_EFM32HG322F64"

**Comment by chrta** [comment]
This is now work in progress. I will write a comment, when i think all comments are addressed.

**Comment by chrta** [comment]
@galak 

> Can you add a README per: http://docs.zephyrproject.org/contribute/contribute_non-apache.html
> for the SDK import.

A fixed gecko sdk readme is already part of a not yet merge pull request, see commit 
https://github.com/GilBenkoe/zephyr/commit/7bf473905cb2f11b9a3f3fba2ba9d40949a66953
Is this sufficient?

File: boards/arm/efm32hg_slstk3400a/doc/efm32hg_slstk3400a.rst [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+.. code-block:: console
+
+   Hello World! arm
+
+
+.. _EFM32HG-SLSTK3400A Website:
+   https://www.silabs.com/products/development-tools/mcu/32-bit/efm32-happy-gecko-starter-kit
+
+.. _EFM32HG-SLSTK3400A User Guide:
+   http://www.silabs.com/documents/public/user-guides/efm32wg-stk3800-ug.pdf
```
Comment (chrta) [inline boards/arm/efm32hg_slstk3400a/doc/efm32hg_slstk3400a.rst]: Some links here are wrong.

File: drivers/serial/uart_gecko.c (hunk 9 above)
Comment (chrta) [reply drivers/serial/uart_gecko.c]: I removed the this from the serial driver and reconfigured the board and soc to use usart1 instead of uart1.

**Comment by chrta** [comment]
Every comment should be addressed now (only the README for non-apache-projects is not added).
Can someone point be to the issue that shippable found? I am not able to find the reports where i can see which test failed.

**Comment by chrta** [comment]
The tests that are failing overflow the sram (only 8kB), e.g.:
~~~
./scripts/sanitycheck -p efm32hg_slstk3400a -s tests/posix/pthread_join/portability.posix
~~~

Result:
~~~
[ 94%] Linking C executable zephyr_prebuilt.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:       23768 B        64 KB     36.27%
            SRAM:        9572 B         8 KB    116.85%
        IDT_LIST:          40 B         2 KB      1.95%
/opt/zephyr-sdk/sysroots/x86_64-pokysdk-linux/usr/libexec/arm-zephyr-eabi/gcc/arm-zephyr-eabi/6.2.0/real-ld: zephyr_prebuilt.elf section `noinit' will not fit in region `SRAM'
/opt/zephyr-sdk/sysroots/x86_64-pokysdk-linux/usr/libexec/arm-zephyr-eabi/gcc/arm-zephyr-eabi/6.2.0/real-ld: section .intList VMA [0000000020002000,0000000020002027] overlaps section noinit VMA [0000000020000548,00000000200021c7]
/opt/zephyr-sdk/sysroots/x86_64-pokysdk-linux/usr/libexec/arm-zephyr-eabi/gcc/arm-zephyr-eabi/6.2.0/real-ld: region `SRAM' overflowed by 1380 bytes
collect2: error: ld returned 1 exit status
~~~

On the ST board nucleo_f030r8 (also 8kB) it fits:
~~~
SRAM:        7672 B         8 KB     93.65%
~~~

What should i do to fix this?

**Comment by diegosueiro** [comment]
@chrta,

> What should I do to fix this?

This is what I did to fix the flash overflow. You can follow the same approach for the sram:
https://github.com/zephyrproject-rtos/zephyr/pull/7008/commits/9858d8e4ed7564e1c7afe5c95ddd2c2ff5b4af74

I think you will have to create another PR just to fix this.

**Comment by chrta** [comment]
@diegosueiro

>> What should I do to fix this?

> This is what I did to fix the flash overflow. You can follow the same approach for the sram: 9858d8e

These tests and samples link for an stm32 with 8kb ram, but do not link for this efm32 with 8kb ram. I assume the reason is the size of the gecko sdk code pulled into the project. I do not want to raise the ram limit, just because the gecko sdk eats so much of it.
Any other suggestion?

**Comment by chrta** [comment]
@nashif I think something went wrong here. This is marked as merged, but the commits are not in master. And i think that should not be merged right now, since i am currently fixing the sanitycheck issues.

**Comment by diegosueiro** [comment]
@chrta,

If you don't want to change the test requirements the only thing I can think about to overcome this is to remove the GPIO and UART enablement in the default configuration of the board.

PS.: It is showing that this PR has 0 commits. This is really strange.

**Comment by chrta** [comment]
@diegosueiro By reducing the stack sizes all tests are working now. So no need to change anything else.

@nashif @galak  Could you please repoen this pull request?

**Comment by galak** [comment]
> @nashif @galak Could you please reopen this pull request?

reopen how?  looks like we merged a version of this.

**Comment by galak** [comment]
> reopen how? looks like we merged a version of this.

or do you mean #6078 ?

**Comment by chrta** [comment]
> or do you mean #6078 ?

No, i mean this one. There is no commit merged from this pull request. The status of this merge request indicates this, but if i look into the commits of the master branch, i do not see any commit from this pull request.

If reopening is not possible, i can just open a new pull request and reference this one from the new request.

**Comment by galak** [comment]
> If reopening is not possible, i can just open a new pull request and reference this one from the new request.

Probably just open a new PR.

**Pull request merged**